rapid7/metasploit-framework
# AWS SSM Sessions

Amazon Web Services provides conveniently privileged backdoors in the form of their SSM agents which do not require connectivity with the target instance, merely valid credentials to AWS' API. Due to this indirect "connection" paradigm, this mechanism can be used to control otherwise "air-gapped" targets.

At the user-facing level, this PR provides two abstractions to the session handler atop which it can operate a command session.

1. Script-execution abstraction and pickup via AWS' API. This approach abstracts asynchronous request/response parsing for SSM requests into an IO channel with which the AWS SSM client is then wrapped to emulate the expected Stream. The mechanism is rather raw and could use better error handling, retries on laggy output, and a threadsafe cursor implementation.
There is a significant limitation with these sessions not present in normal stream-wise abstractions: a response limit of 2500 chars. This limitation can be overcome by utilizing an S3 bucket to store command output; however, due to the nature of access we seek to obtain, it would not only add to the logged event loads but retain the results of our TTPs in a "buffer" accessible to other people. This functionality can be added down the line in the form of S3 config options in the handler to be passed into the SSM client for command execution and acquisition of output.

2. WebSocket-based "direct" interaction using @zeroSteiner's marvelous Rex implementation for WebSocket channels providing the necessary hooks by which to effect relevant encoding and decoding of "higher-level" protocols. This modality is more "mature" (though somewhat less novel than the C2 abstraction) in terms of interface presented, and supports fully interactive, colorful sessions via the stream abstraction.

**Testing:**
  Gets sessions, provides command IO, leaves a bunch of log entries
in CloudTrail (something to keep in mind for opsec considerations).

**How to test:**
1. Pull PR into your test framework and install the resulting bundle to pull in the SSM SDK
2. Start the updated console
3. `use auxiliary/cloud/aws/enum_ssm`
4. `set SECRET_ACCESS_KEY` to your secret key
5. `set ACCESS_KEY_ID` to your access key ID
6. `set CreateSession false`
8. `exploit`
9. verify how many and which systems you're about to shell (this thing is an autopwn on roids in the wrong environment)
10. `set CreateSession true` if there are valid targets and they are safe to shell (over TLS, but still worth considering)
11. `exploit`

Advanced users can test the payloads (unix^windows) directly via the handler and setting the provided datastore options.

Notes:
* In my testing, running `exploit` just keeps on getting more shells. Not sure if its related to my mutant framework setup, but worth testing for the effect upstream (and figuring out why it happens, if it does).
* Suggested mitigation for this threat vector (OOB privileged access permitting command and raw IO) is to **uninstall the bloody agent and make sure it stays that way** (or just build and use images which have never had it on them in the first place, nix `cloud-init` after boostrapping as well, while at it)
* The two session abstraction stacks this work presents can (and IMO should) be used to build C2 sessions and tramsports to permit comms with "third-party software" (among other things) :wink:
